### ISSUE TITLE
Option to build third party dependencies instead of requiring them in cmake ?

### ISSUE THREAD

**Issue by nemerle** [BearishSun/BansheeEngine#98]
At least some of the Banshee's dependencies are not commonly packaged on linux ( nvtt,PhysX,fbxsdk ), also there is  another _interesting_ issue related to project building, namely:

What's going to happen when a system-provided dependency has older/newer version than the project needs ?

IMHO building the project on Linux is going to either require a thorough documentation for various distributions, or some amount of CMake's **external_project** magic :smile: 

While pre-compiled dependencies are great for saving on build times, they sadly become somewhat of an obstacle when dealing with cross-platform projects :cry:

**Comment by BearishSun** [comment]
Banshee looks for dependencies locally first, so as long as the pre-compiled dependencies are present the system provided ones should not be relevant.

**Comment by nemerle** [comment]
Well I don't think that'll work very well linux  ( precompiled dependencies ). there are some issues with compiler abi compability across distributions ( or maybe it's just my trauma from gcc 5.1 and that std::string thingy :) )

**Comment by BearishSun** [comment]
I don't think there should be any ABI issues. Library like FBX SDK doesn't even come with source code so they must be handling it somehow.

**Comment by nemerle** [comment]
If it's static library, then yup, should work ( if there are no trouble with std::string/std::__cxx11::string issues ), dynamic libraries might be painful though :/

**Comment by BearishSun** [comment]
In any case, it shouldn't be too difficult to set up the build system to automatically fetch the dependency source and build it. There is already some functionality for it there, as I use it for often-changing dependencies (it's easier to recompile them on-the-fly than to keep reuploading the binary packages).

**Comment by nemerle** [comment]
Yup, I'm doing that in a few of my own projects as well ( [example here](https://github.com/Segs/Segs/blob/master/3rd_party/CMakeLists.txt) )

**Comment by slouffka** [comment]
Hey. I'm actually trying to build Banshee3D under Arch Linux. And looks like the only dep I'm missing is nvtt. I just want to give the engine a try but the CMake module for nvtt searches it in some Dependencies dir which I even don't see in this repo. How can I update that module so it would use nvtt.h from /usr/include/nvtt? +)

**Comment by slouffka** [comment]
I tried to change nvtt install dir this way. Doing something wrong?
```
# set(nvtt_INSTALL_DIR ${PROJECT_SOURCE_DIR}/../Dependencies/nvtt CACHE PATH "")
set(nvtt_INSTALL_DIR /usr/includes/nvtt CACHE PATH "")                          
gen_default_lib_search_dirs(nvtt)  
```

**Comment by slouffka** [comment]
Looks like I need not use gen_default_lib_search_dirs. I'll try to find workaround.

**Comment by BearishSun** [comment]
Are you compiling nvtt from https://github.com/BearishSun/nvidia-texture-tools or from the official repo?

Make sure to compile from the above fork instead of the official repo, as it has install paths changed (not only changed, but fixed, since it was broken on the official repo).

Banshee looks for nvtt in the following paths:
*/include/nvtt.h
*/lib/libbc6h.a
*/lib/libbc7.a
*/lib/libnvcore.a
*/lib/libnvimage.a
*/lib/libnvmath.a
*/lib/libnvthread.a
*/lib/libnvtt.a
*/lib/libsquish.a

Where * are different paths, searched in order:
BansheeRoot/Dependencies/nvtt
/usr/local/
...other global system paths

The offical nvtt build places nvtt.h in an extra `nvtt` subfolder, and the libraries in an extra `static` subfolder (it also fails to install a couple of libraries).

**Comment by slouffka** [comment]
Nope. Using this one https://github.com/castano/nvidia-texture-tools

**Comment by BearishSun** [comment]
That's the issue then, compile the one above and it should work.

**Comment by slouffka** [comment]
It compiles debug version but you've mentioned the ones witout d_

```
➜  nvidia-texture-tools git:(master) ✗ l /usr/local/lib                         <<<
итого 587M
drwxr-xr-x  3 root root 4,0K окт 19 14:47 .
drwxr-xr-x 12 root root 4,0K окт 15 22:17 ..
-rw-r--r--  1 root root 202K окт 19 14:45 libbc6h_d.a
-rw-r--r--  1 root root 790K окт 19 14:44 libbc7_d.a
lrwxrwxrwx  1 root root   15 мар  6  2017 libentityx.so -> libentityx.so.1
lrwxrwxrwx  1 root root   19 мар  6  2017 libentityx.so.1 -> libentityx.so.1.1.2
-rwxr-xr-x  1 root root 483K мар  6  2017 libentityx.so.1.1.2
-rwxr-xr-x  1 root root 371M мар 10  2017 libjs_static.ajs
-rwxr-xr-x  1 root root 212M мар 10  2017 libmozjs-55a1.so
-rw-r--r--  1 root root 152K окт 19 14:44 libnvcore_d.a
-rw-r--r--  1 root root 944K окт 19 14:45 libnvimage_d.a
-rw-r--r--  1 root root 260K окт 19 14:44 libnvmath_d.a
-rw-r--r--  1 root root 114K окт 19 14:45 libnvthread_d.a
-rw-r--r--  1 root root 1,5M окт 19 14:45 libnvtt_d.a
-rw-r--r--  1 root root 235K окт 19 14:45 libsquish_d.a
drwxr-xr-x  2 root root 4,0K мар 10  2017 pkgconfig
```

Do you know how to make release version? I can't see release target. Sadly I'm not too advanced in make.

**Comment by BearishSun** [comment]
You need to set the CMake variable "CMAKE_BUILD_TYPE" to "Release". You can do it using the CMake command line (`-D name=value`), or just type it into the GUI if using CMake GUI.

Also please not that Banshee Linux port is not yet complete, you should be able to compile it but running it will not work at the moment.

**Comment by slouffka** [comment]
I understand that it could not work as expected or don't work at all. But I'm very curious if it will. Thank you for your help. You're the most supportive person on Github I've ever seen ;) Keep up the good work.

**Comment by slouffka** [comment]
Actually it's cmake -DCMAKE_BUILD_TYPE=Release

**Comment by slouffka** [comment]
Okay. It worked. Next step is PhysX.

**Comment by BearishSun** [comment]
There's no simple way to disable these features, but they could be removed with varying amount of effort:
 - FBX is only needed for the FBX importer plugin. You can just remove it from the build, or skip building that target, and then tell the application not to load that specific plugin in `START_UP_DESC`.
 - nvtt is only used in `BsPixelUtil.cpp`. You could provide different implementations for the relevant functionality, without nvtt. Or just provide dummy methods if you are sure you are never going to call them.
 - PhysX is the hardest to remove, but it is also a plugin with a clearly defined API and interchangeable implementation. You'd need to create another physics plugin that either uses some other physics API, or does nothing (e.g. `NullPhysics`).

**Comment by Anth0rx** [comment]
Okay, so to remove FBX support seems to be rather simple.
Regarding nvtt I will have a look at the corresponding file.
Why are you not using bullet instead of PhysX?

**Comment by BearishSun** [comment]
It seemed like a safer bet regarding features/performance, considering big commercial engines use it. But I also made the physics system plugin based so we can later integrate bullet or Havok if needed.
